Classification/FeatureSelection/src/main/java/org/tribuo/classification/fs/FSMatrix.java

```diff
@@ -0,0 +1,191 @@
+     * @return The feature map.
+     */
+    ImmutableFeatureMap getFeatureMap();
+
+    /**
+     * Calculates the mutual information I(X_i;Y) where Y is the label
+     * @param featureIndex The index i
+     * @return I(X_i;Y)
+     */
+    double mi(int featureIndex);
+
+    /**
+     * Calculates the mutual information I(X_i;X_j).
+     * @param firstIndex The index i
+     * @param secondIndex The index j
+     * @return I(X_i;X_j)
+     */
+    double mi(int firstIndex, int secondIndex);
+
+    /**
+     * Generates an array of I(X_i;Y), ranging X_i over all possible features, where Y is the label.
+     * @return An array where array[i] = I(X_i;Y)
+     */
+    default double[] miList() {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = mi(i);
+        }
+        return row;
+    }
+
+    /**
+     * Generates an array of I(X_i;X_j), ranging X_i over all possible features.
+     * @param targetIndex The index of X_j variable
+     * @return An array where array[i] = I(X_i;X_j)
+     */
+    default double[] miList(int targetIndex) {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = mi(i,targetIndex);
+        }
+        return row;
+    }
+
+    /**
+     * Calculates the joint mutual information I(X_i,X_j;Y) where Y is the label
+     * @param featureIndex The index i
+     * @param jointIndex The index j
+     * @return I(X_iX_j;Y)
+     */
+    double jmi(int featureIndex, int jointIndex);
+
+    /**
+     * Calculates the joint mutual information I(X_iX_j;X_k)
+     * @param firstIndex The index i
+     * @param jointIndex The index j
+     * @param targetIndex The index k
+     * @return I(X_iX_j;X_k)
+     */
+    double jmi(int firstIndex, int jointIndex, int targetIndex);
+
+    /**
+     * Generates an array of I(X_iZ;Y), ranging X_i over all possible features, where Y is the label.
+     * @param jointIndex The index of the Z variable
+     * @return An array where array[i] = I(X_iZ;Y)
+     */
+    default double[] jmiList(int jointIndex) {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = jmi(i,jointIndex);
+        }
+        return row;
+    }
+
+    /**
+     * Generates an array of I(X_iZ;Y), ranging X_i over all possible features.
+     * @param targetIndex The index of the Y variable
+     * @param jointIndex The index of the Z variable
+     * @return An array where array[i] = I(X_iZ;Y)
+     */
+    default double[] jmiList(int targetIndex, int jointIndex) {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = jmi(i,jointIndex,targetIndex);
```

Comment:
Why does this switch the argument order and naming from the previous version?
```suggestion
     * Generates an array of I(X_i,X_j;Y), ranging X_i over all possible features.
     * @param secondIndex The index of the X_j variable
     * @param targetIndex The index of the X_k variable
     * @return An array where array[i] = I(X_i,X_j;X_k)
     */
    default double[] jmiList(int secondIndex, int targetIndex) {
        int numFeatures = getNumFeatures();
        double[] row = new double[numFeatures];
        for (int i = 0; i < numFeatures; i++) {
            row[i] = jmi(i,secondIndex,targetIndex);
```